### ISSUE TITLE
JSON transformer behaves incorrectly when data has `id` field

### ISSUE THREAD

**Issue by iv-mexx** [gatsbyjs/gatsby#2210]
Gatsby Version: 1.1.6
Node.js Version: v8.2.1
OS: macOS Sierra 10.12.6

I've created a [Demo Project](https://github.com/iv-mexx/gatsby-issue-demo)

---

When loading JSON (probably other serialisers as well), nodes of the JSON automatically are assigned an `id` attribute which seems to be the absolute file path + some additional information.

If the JSON data contains an `id` attribute itself, things break.

In the linked sample project, there is a file `fruits.json` that looks like this:

<details>
<summary>fruits.json</summary><p>

```
[
  {
    "id": "0",
    "name": "Apple"
  },
  {
    "id": "1",
    "name": "Orange"
  },
  {
    "id": "2",
    "name": "Lemon"
  }
]
```

</p></details>

Now, if a second file (`cars.json`) is added, keys start to conflict and overwrite each other. 

<details>
<summary>cars.json</summary><p>

```
[
  {
    "id": "0",
    "make": "Toyota"
  },
  {
    "id": "1",
    "make": "BMW"
  },
  {
    "id": "3",
    "make": "Skoda"
  }
]
```

</p></details>

E.g in this case, the query returns the following result. Notably, cars with `id` 0 and 1 are missing. If the car "Skoda" had `id` = 2, there would be no `allCarsJson` at all!

<details>
<summary>Query</summary><p>

```

{
  allFruitsJson {
    edges {
      node {
        id
        name
      }
    }
  }
  allCarsJson {
    edges {
      node {
        id
        make
      }
    }
  }
}
```

</p></details>

<details>
<summary>Results</summary><p>

```
{
  "data": {
    "allFruitsJson": {
      "edges": [
        {
          "node": {
            "id": "0",
            "name": "Apple"
          }
        },
        {
          "node": {
            "id": "1",
            "name": "Orange"
          }
        },
        {
          "node": {
            "id": "2",
            "name": "Lemon"
          }
        }
      ]
    },
    "allCarsJson": {
      "edges": [
        {
          "node": {
            "id": "3",
            "make": "Skoda"
          }
        }
      ]
    }
  }
}
```

</p></details>

---

Btw, if the json contains `parent` and `childs` attributes, they are overwritten by Gatsby

I'm not sure what the correct behaviour here should be, e.g. if `id` attributes in json should be forbidden, if gatsby should overwrite the attribute, or if there is another way to solve this...
*At the very least hover*, Gatsby should currently generate error messages to warn about this!

**Comment by KyleAMathews** [comment]
Yeah… Gatsby IDs need to be globally unique. Which makes it tough when there can be user supplied IDs. Possibly the answer is to automatically "globalize" your IDs. Or perhaps just never use user-supplied IDs and instead copy the supplied ID to something like `_id` and generate a GUID.

**Comment by swim** [comment]
Maybe some warnings during compile could prevent some head scratching RE parent and childs keys. Where would you envision the UUID (globalizer) lives? Could this be added as a standalone transformer?

**Comment by CanRau** [comment]
maybe that's what you meant, but how about gatsby always uses _id, which might be less likely to conflict with user provided ids and in case someone still adds custom _id throw a warning

**Comment by hotzeplotz** [comment]
i have been struggling with this for a while until i instrumented `gatsby-node.js` in `gatsby-transformer-json` and traced the issue to `createNode()`, whose documentation in fact clearly states that `id` needs to be globally unique (https://www.gatsbyjs.org/docs/bound-action-creators/#createNode).

i agree however that it would be helpful to at the very least let Gatsby output a warning when a node is discarded: at a quick glance this could be done within `createNode()` itself at https://github.com/gatsbyjs/gatsby/blob/e1bebad2932a333322a0c7c98cd0d7ff367f1c8a/packages/gatsby/src/redux/actions.js#L536 - am i looking at the right place to attempt to add some informative output for this scenario?

**Comment by pieh** [comment]
@hotzeplotz This is not not simple unfortunately - gatsby persist nodes between builds, so it know that it doesn't need to rerun queries if nodes that were used in queries didn't change. So gatsby will very frequently override already existing nodes. This will also be problematic in `develop` mode when user can update files (f.e. markdown) and updated nodes will again override previous ones.
